### PR TITLE
[Serve] Honor X-Request-ID for setting request id

### PR DESCRIPTION
**Pull request by sihanwang41**
## Why are these changes needed?

- honor `X-Request-ID` in http request.
- all request id will be coming from the middleware instead of generating the request handle code path.
- use uuid4 as request id by default.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by sihanwang41** [comment]
@edoakes @GeneDer @zcin @shrekris-anyscale ping for review :)

File: python/ray/serve/_private/http_proxy.py [hunk 1]
```diff
@@ -482,7 +483,8 @@ async def __call__(self, scope, receive, send):
                 scope["path"] = route_path.replace(route_prefix, "", 1)
                 scope["root_path"] = root_path + route_prefix
 
-            request_id = get_random_letters(10)
+            # RequestIdMiddleware makes sure scoep has x-request-id attribute
+            request_id = MutableHeaders(scope=scope).get("x-request-id")
             request_context_info = {
                 "route": route_path,
                 "request_id": request_id,
```
Comment (GeneDer) [inline python/ray/serve/_private/http_proxy.py:490]: Should we also change line 499 to respect this? (basically asking if "x-request-id" should take precedent)

File: python/ray/serve/_private/http_proxy.py [hunk 2]
```diff
@@ -857,13 +859,29 @@ def __init__(self, app):
...
-            request_id = ray.serve.context._serve_request_context.get().request_id
             if message["type"] == "http.response.start":
                 headers = MutableHeaders(scope=message)
+                # TODO(Sihan) Deprecate RAY_SERVE_REQUEST_ID_HEADER
                 headers.append(RAY_SERVE_REQUEST_ID_HEADER, request_id)
+                headers.append("X-Request-ID", request_id)
             if message["type"] == "websocket.accept":
+                # TODO(Sihan) Deprecate RAY_SERVE_REQUEST_ID_HEADER
                 message[RAY_SERVE_REQUEST_ID_HEADER] = request_id
+                message["X-Request-ID"] = request_id
```
Comment (GeneDer) [inline python/ray/serve/_private/http_proxy.py:884]: For these "X-Request-ID" do we need to all cap instead?

File: python/ray/serve/tests/test_http_headers.py [hunk 3]
```diff
@@ -22,17 +23,28 @@ def __call__(self):
...
     def verify_result(self):
-        resp = requests.get(
-            "http://localhost:8000", headers={RAY_SERVE_REQUEST_ID_HEADER: "123-234"}
-        )
-        assert resp.status_code == 200
-        assert resp.json() == 1
-        assert RAY_SERVE_REQUEST_ID_HEADER in resp.headers
-        assert resp.headers[RAY_SERVE_REQUEST_ID_HEADER] == "123-234"
+
+        for header_attr in [RAY_SERVE_REQUEST_ID_HEADER, "X-Request-ID"]:
```
Comment (GeneDer) [inline python/ray/serve/tests/test_http_headers.py:41]: Would be nice if we can add cases for when both are set at the same time with different values :)

File: python/ray/serve/_private/http_proxy.py (hunk 2 above)
Comment (sihanwang41) [reply python/ray/serve/_private/http_proxy.py:884]: ditto

File: python/ray/serve/_private/http_proxy.py [hunk 4]
```diff
@@ -482,7 +483,8 @@ async def __call__(self, scope, receive, send):
                 scope["path"] = route_path.replace(route_prefix, "", 1)
                 scope["root_path"] = root_path + route_prefix
 
-            request_id = get_random_letters(10)
+            # RequestIdMiddleware makes sure scoep has x-request-id attribute
```
Comment (edoakes) [inline python/ray/serve/_private/http_proxy.py:486]: typo and fix punctuation

File: python/ray/serve/_private/constants.py [hunk 5]
```diff
@@ -200,7 +200,7 @@
 )
 
 # Request ID used for logging. Can be provided as a request
-# header and will always be returned as a response header.
+# header and will always be returned as a response header.(deprecated)
```
Comment (edoakes) [inline python/ray/serve/_private/constants.py:203]:
```suggestion
# header and will always be returned as a response header.
# DEPRECATED: use `X-Request-Id` instead.
```

File: python/ray/serve/_private/http_proxy.py [hunk 6]
```diff
@@ -857,13 +859,29 @@ def __init__(self, app):
         self.app = app
 
     async def __call__(self, scope, receive, send):
+
+        headers = MutableHeaders(scope=scope)
+        if RAY_SERVE_REQUEST_ID_HEADER not in headers and "x-request-id" not in headers:
+            # If X-Request-ID and RAY_SERVE_REQUEST_ID_HEADER are both not set, we
+            # generate a new request ID.
+            request_id = str(uuid.uuid4())
+            headers.append("x-request-id", request_id)
```
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/_private/http_proxy.py:868]: for now we should also populate `ray_serve_request_id` here too right?

**Review by edoakes — COMMENTED** [review]
@sihanwang41 please update the docs you added to use `X-Request-Id` instead

File: python/ray/serve/_private/http_proxy.py (hunk 2 above)
Comment (GeneDer) [reply python/ray/serve/_private/http_proxy.py:884]: Similarly, as long as we pick a specific spelling, it's totally fine by me :)

**Review by GeneDer — APPROVED** [review]
LGTM

File: python/ray/serve/tests/test_http_headers.py [hunk 7]
```diff
@@ -72,6 +84,31 @@ def __call__(self) -> int:
...
+        "http://localhost:8000",
+        headers={
+            RAY_SERVE_REQUEST_ID_HEADER: "123",
+            "X-Request-ID": "234",
+        },
+    )
+
+    assert resp.status_code == 200
+    assert "x-request-id" in resp.headers
+    assert "234" == resp.headers["x-request-id"]
```
Comment (GeneDer) [inline python/ray/serve/tests/test_http_headers.py:109]: Actually can we check `resp.text` as well just to double check?

**Review by edoakes — APPROVED** [review]

**Comment by sihanwang41** [comment]
@edoakes ping for merge, failure unrelated.

**Pull request merged**